spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RecurrentDecoder.scala

```diff
@@ -41,26 +42,84 @@ class RecurrentDecoder[T : ClassTag](seqLength: Int)
   (implicit ev: TensorNumeric[T]) extends Recurrent[T] {
 
   times = seqLength
-  private val newInput = Tensor[T]()
+
+  // states, gradStates only used with multirnncell and key starts from 0
+  private val states: Table = T()
+  private val gradStates = T()
+
+  override def getStates(): Activity = {
+    if (containMultiRNNCell) {
+      states
+    } else super.getStates()
+  }
+
+  override def initHidden(sizes: Array[Int]): Unit = {
+    val stepSizes = sizes
+
+    if (containMultiRNNCell) {
+      if (states.getState().size == 0) {
+        cells.clear()
+        cells += topology
+      }
+
+      val multiCells = topology.asInstanceOf[MultiRNNCell[T]].cells
+      var i = 0
+      while (i < multiCells.size) {
+        states(i) = multiCells(i).hidResize(null, batchSize, stepSizes)
+        gradStates(i) = multiCells(i).hidResize(null, batchSize, stepSizes)
+        i += 1
+      }
+    } else super.initHidden(sizes)
+  }
+
+  /**
+   *
+   *  modules: -- preTopology
+   *           |- topology (cell)
+   *
+   * The topology (or cell) will be cloned for N times w.r.t the time dimension.
+   * The preTopology will be execute only once before the recurrence.
+   *
+   * @param module module to be add
+   * @return this container
+   */
+  override def add(module: AbstractModule[_ <: Activity, _ <: Activity, T]):
+    RecurrentDecoder.this.type = {
+    require(module.isInstanceOf[Cell[T]],
+      "Recurrent: contained module should be Cell type")
+
+    topology = module.asInstanceOf[Cell[T]]
+    preTopology = topology.preTopology
+
+    if (preTopology != null) {
+      modules += preTopology
+    }
+    modules += topology
+
+    require((preTopology == null && modules.length == 1) ||
+      (topology != null && preTopology != null && modules.length == 2),
+      "Recurrent extend: should contain only one cell or plus a pre-topology" +
+        " to process input")
+    this
+  }
 
   override def updateOutput(input: Tensor[T]): Tensor[T] = {
     require(input.dim == 2 || input.dim == 4 || input.dim == 5,
       "Recurrent: input should be a 2D/4D/5D Tensor, e.g [batch, nDim], " +
         s"current input.dim = ${input.dim}")
-
-    batchSize = input.size(batchDim)
-
     val hiddenSize = topology.hiddensShape(0)
     val outputSize = input.size()
+    // output size is batch x hidden x featureSize
     outputSize(1) = hiddenSize
-    require(hiddenSize == input.size()(1), "hiddenSize is " +
-      "not the same with input size!! Please update cell settings or use Recurrent instead!")
+    batchSize = input.size(batchDim)
     val featureSizes = outputSize.drop(1)
     output.resize(Array(batchSize, times) ++ featureSizes)
-    // Clone N modules along the sequence dimension.
-    extend(featureSizes)
-    if (preTopology != null) newInput.resize(output.size())
-    else outputCell.resize(output.size())
+    if (preTopology != null) {
+      val sizes = output.size()
+      sizes(2) = hiddenSize * 4
```

Comment:
why hiddenSize * 4?

Reply:
The preTopology gathered all linear computing with input. and there is 4 linears (i2g, f2g, c2g, o2g). So the expected hiddensize after gathering is hiddenSize * 4

Reply:
you cannot assume LSTM here; what about RNN, GRU, etc?